### PR TITLE
 Jira Source HTTP CDK

### LINKED ISSUE #3453: Source Jira: update to use CDK
Rewriting Jira Source to use HTTP CDK

### PR DESCRIPTION
**Pull request by vitaliizazmic**
## What
Migrate Jira Source to HTTP CDK

Closes #3453

### CODE REVIEW

File: airbyte-integrations/connectors/source-jira/source_jira/api.py [hunk 1]
```diff
@@ -0,0 +1,590 @@
...
+    def path(self, stream_slice: Mapping[str, Any] = None, **kwargs) -> str:
+        key = stream_slice["key"]
+        return f"project/{key}/securitylevel"
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        response_json = response.json()
+        records = [
+            response_json,
+        ]
+        yield from records
```
Comment (keu) [inline airbyte-integrations/connectors/source-jira/source_jira/api.py:490]: just `yield records` ?

**Review by keu — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-jira/source_jira/source.py [hunk 2]
```diff
@@ -22,11 +22,144 @@
...
+        error_msg = None
+
+        try:
+            authenticator = self.get_authenticator(config)
+            args = {"authenticator": authenticator, "domain": config["domain"]}
+            issue_resolutions = IssueResolutions(**args)
+            for item in issue_resolutions.read_records(sync_mode=SyncMode.full_refresh):
+                continue
+        except ConnectionError as error:
+            alive, error_msg = False, str(error)
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-jira/source_jira/source.py:103]:
```suggestion
            alive, error_msg = False, repr(error)
```

File: airbyte-integrations/connectors/source-jira/setup.py [hunk 3]
```diff
@@ -25,10 +25,11 @@
 
 from setuptools import find_packages, setup
 
-MAIN_REQUIREMENTS = ["airbyte-cdk", "requests", "pendulum==2.1.2"]
+MAIN_REQUIREMENTS = ["airbyte-cdk", "requests", "pendulum"]
```
Comment (keu) [inline airbyte-integrations/connectors/source-jira/setup.py:28]: pin cdk~=0.1

**Review by keu — APPROVED** [review]
just small comment

File: airbyte-integrations/connectors/source-jira/source_jira/streams.py [hunk 4]
```diff
@@ -0,0 +1,570 @@
...
+API_VERSION = 3
+
+
+class JiraStream(HttpStream, ABC):
+    """
+    Jira API Reference: https://developer.atlassian.com/cloud/jira/platform/rest/v3/intro/
+    """
+
+    primary_key = "id"
+    parse_response_root = ""
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-jira/source_jira/streams.py:44]: is this better than empty stirng? 

```suggestion
    parse_response_root = None
```

File: airbyte-integrations/connectors/source-jira/source_jira/streams.py [hunk 5]
```diff
@@ -0,0 +1,570 @@
...
+
+class ScreenTabs(JiraStream):
+    def path(self, stream_slice: Mapping[str, Any] = None, **kwargs) -> str:
+        screen_id = stream_slice["screen_id"]
+        return f"screens/{screen_id}/tabs"
+
+    def read_records(self, stream_slice: Optional[Mapping[str, Any]] = None, **kwargs) -> Iterable[Mapping[str, Any]]:
+        screens_stream = Screens(authenticator=self.authenticator, domain=self._domain)
+        for screen in screens_stream.read_records(sync_mode=SyncMode.full_refresh):
+            yield from super().read_records(stream_slice={"screen_id": screen["id"]}, **kwargs)
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-jira/source_jira/streams.py:511]: should this just call `read_tab_records`?

File: airbyte-integrations/connectors/source-jira/source_jira/streams.py [hunk 6]
```diff
@@ -0,0 +1,570 @@
...
+            yield from super().read_records(stream_slice={"key": project["key"]}, **kwargs)
+
+
+class ProjectEmail(JiraStream):
+    def path(self, stream_slice: Mapping[str, Any] = None, **kwargs) -> str:
+        project_id = stream_slice["project_id"]
+        return f"project/{project_id}/email"
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        response_json = response.json()
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-jira/source_jira/streams.py:451]: we can just `yield response.json()` here right?

File: airbyte-integrations/connectors/source-jira/source_jira/streams.py [hunk 7]
```diff
@@ -0,0 +1,570 @@
...
+        return params
+
+    def request_headers(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> Mapping[str, Any]:
+        return {"Accept": "application/json"}
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        response_json = response.json()
+        records = response_json if not self.parse_response_root else response_json.get(self.parse_response_root, [])
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-jira/source_jira/streams.py:91]: is `response_json` guaranteed to be a list?

File: airbyte-integrations/connectors/source-jira/source_jira/streams.py [hunk 8]
```diff
@@ -0,0 +1,570 @@
...
+        stream_state = stream_state or {}
+        params = super().request_params(stream_state=stream_state, **kwargs)
+        params["fields"] = ["attachment", "issuelinks", "security", "issuetype", "created"]
+        if stream_state.get(self.cursor_field):
+            issues_state = pendulum.parse(stream_state.get(self.cursor_field))
+            issues_state_row = issues_state.strftime("%Y/%m/%d %H:%M")
+            params["jql"] = f"created > '{issues_state_row}'"
+        return params
+
+    def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]) -> Mapping[str, Any]:
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-jira/source_jira/streams.py:173]: should this method instead be in the superclass?

File: airbyte-integrations/connectors/source-jira/source_jira/streams.py [hunk 9]
```diff
@@ -0,0 +1,570 @@
...
+        key = stream_slice["key"]
+        return f"issue/{key}/watchers"
+
+    def read_records(self, stream_slice: Optional[Mapping[str, Any]] = None, **kwargs) -> Iterable[Mapping[str, Any]]:
+        issues_stream = Issues(authenticator=self.authenticator, domain=self._domain)
+        for issue in issues_stream.read_records(sync_mode=SyncMode.full_refresh):
+            yield from super().read_records(stream_slice={"key": issue["key"]}, **kwargs)
+
+
+class IssueWorklogs(JiraStream):
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-jira/source_jira/streams.py:339]: should this be incremental?

**Review by sherifnada — CHANGES_REQUESTED** [review]
mostly LGTM with a few questions

File: airbyte-integrations/connectors/source-jira/source_jira/streams.py [hunk 10]
```diff
@@ -0,0 +1,561 @@
...
+    @property
+    @abstractmethod
+    def cursor_field(self) -> str:
+        """
+        Defining a cursor field indicates that a stream is incremental, so any incremental stream must extend this class
+        and define a cursor field.
+        """
+        pass
+
+    def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]) -> Mapping[str, Any]:
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-jira/source_jira/streams.py:108]:
```suggestion
    @abstractmethod
    def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]) -> Mapping[str, Any]:
```

**Review by sherifnada — APPROVED** [review]
One suggestion but otherwise LGTM

File: airbyte-integrations/bases/standard-source-test/src/main/java/io/airbyte/integrations/standardtest/source/StandardSourceTest.java [hunk 11]
```diff
@@ -118,7 +118,8 @@
       "airbyte/source-gitlab-singer",
       "airbyte/source-google-workspace-admin-reports",
       "airbyte/source-zendesk-talk",
-      "airbyte/source-quickbooks-singer");
+      "airbyte/source-quickbooks-singer",
+      "airbyte/source-jira");
```
Comment (sherifnada) [inline airbyte-integrations/bases/standard-source-test/src/main/java/io/airbyte/integrations/standardtest/source/StandardSourceTest.java:122]: if we are using new test script then we don't need this or the build block in gradle

**Pull request merged**